spring-integration-core/src/main/java/org/springframework/integration/router/AbstractMappingMessageRouter.java

```diff
@@ -173,11 +172,13 @@ public void replaceChannelMappings(Properties channelMappings) {
 	}
 
 	private void doSetChannelMappings(Map<String, String> newChannelMappings) {
-		Map<String, String> oldChannelMappings = this.channelMappings;
-		this.channelMappings = newChannelMappings;
+		Map<String, String> oldChannelMappings = new HashMap<String, String>(this.channelMappings);
+		synchronized (this.channelMappings) {
```

Comment:
While `getChannelMappings()` is now only accessed by `RouterMetrics`, I think it needs to get a copy rather than a hard reference - otherwise the map can change while the caller has a reference.